### PR TITLE
feat(compatibility): add mandates support in stripe compatibility

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

This PR enables,

- compatibility layer to setup mandates in hyperswitch core along revoke and execution

- Setup intent route path fix

- Jwe decryption of merchant creds

### Additional Changes

- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?

mandateID
succeded
revoke
mandate_afterRevoke

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 1]
```diff
@@ -10,10 +11,12 @@ use crate::{
     pii::{self, PeekInterface},
     types::{
         api::{admin, enums as api_enums},
-        transformers::{ForeignFrom, ForeignInto},
+        transformers::{ForeignFrom, ForeignInto, ForeignTryFrom, ForeignTryInto},
     },
 };
 
+// pub struct MyMandateData(pub payments::MandateData);
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:18]: Please remove commented code

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 2]
```diff
@@ -143,12 +145,23 @@ pub struct StripePaymentIntentRequest {
...
     fn try_from(item: StripePaymentIntentRequest) -> errors::RouterResult<Self> {
-        Ok(Self {
+        let stripe_mandate = item.payment_method_options.as_ref().and_then(|mandate| {
+            let StripePaymentMethodOptions::Card {
+                request_three_d_secure: _,
+                mandate_options,
+            }: StripePaymentMethodOptions = mandate.clone();
+            mandate_options
+        });
+        let f = (stripe_mandate, item.currency.to_owned());
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:162]: Name the variable appropriately

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 3]
```diff
@@ -296,7 +315,7 @@ pub struct StripePaymentIntentResponse {
     pub shipping: Option<payments::Address>,
     pub billing: Option<payments::Address>,
     #[serde(with = "common_utils::custom_serde::iso8601::option")]
-    pub capture_on: Option<time::PrimitiveDateTime>,
+    pub capture_on: Option<PrimitiveDateTime>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:318]: don't use specific imports. Undo this change.

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 4]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
 pub enum StripePaymentMethodOptions {
     Card {
         request_three_d_secure: Option<Request3DS>,
+        mandate_options: Option<MandateOption>,
     },
 }
 
-#[derive(Default, Eq, PartialEq, Serialize, Deserialize, Clone)]
+#[derive(PartialEq, Eq, Deserialize, Clone, Default, Debug)]
+#[serde(rename_all = "snake_case")]
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:522]: since the fields are already in snake_case this is not required I guess.

Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:522]: since the fields are already in snake_case this is not required I guess.

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 5]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+        (mandate_options, currency): (Option<MandateOption>, Option<String>),
+    ) -> errors::RouterResult<Self> {
+        let curr: Option<api_enums::Currency> = currency
+            .as_ref()
+            .map(|c| c.to_uppercase().parse_enum("currency"))
+            .transpose()
+            .change_context(errors::ApiErrorResponse::InvalidDataValue {
+                field_name: "currency",
+            })?;
+        let extracted_currmatch = match curr {
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:544]: name the variables appropriately

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 6]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+            Some(curr) => curr,
+            None => {
+                return Err(errors::ApiErrorResponse::InvalidDataValue {
+                    field_name: "currency",
+                })
+                .into_report()
+            }
+        };
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:552]:
```suggestion
       let currency = currency
            .ok_or(errors::ApiErrorResponse::MissingRequiredField {
                field_name: "currency",
            })
            .into_report()
            .and_then(|c| {
                c.to_uppercase().parse_enum("currency").change_context(
                    errors::ApiErrorResponse::InvalidDataValue {
                        field_name: "currency",
                    },
                )
            })?;
```

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 7]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+                    mandate_type: if mandate.mandate_type == Some("multi_use".to_string()) {
+                        payments::MandateType::MultiUse(None)
+                    // payments::MandateType::MultiUse(Some(payments::MandateAmountData {
+                    // amount: mandate.amount.unwrap_or_default(),
+                    // currency: extracted_currmatch
+                    // }))
+                    } else {
+                        payments::MandateType::SingleUse(payments::MandateAmountData {
+                            amount: mandate.amount.unwrap_or_default(),
+                            currency: extracted_currmatch,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:565]:
```suggestion
                            currency,
```

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 8]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
 }
 
-#[derive(Default, Eq, PartialEq, Serialize, Deserialize, Clone)]
+#[derive(PartialEq, Eq, Deserialize, Clone, Default, Debug)]
+#[serde(rename_all = "snake_case")]
+pub struct MandateOption {
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub accepted_at: Option<PrimitiveDateTime>,
+    pub user_agent: Option<String>,
+    pub ip_address: Option<String>,
```
Comment (jarnura) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:527]: Make this field as secret

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 9]
```diff
@@ -490,15 +509,78 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+        };
+        Ok({
+            match mandate_options {
+                Some(mandate) => Some(payments::MandateData {
+                    mandate_type: if mandate.mandate_type == Some("multi_use".to_string()) {
+                        payments::MandateType::MultiUse(None)
+                    // payments::MandateType::MultiUse(Some(payments::MandateAmountData {
+                    // amount: mandate.amount.unwrap_or_default(),
+                    // currency: extracted_currmatch
+                    // }))
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:561]: remove commented code

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 10]
```diff
@@ -143,12 +145,23 @@ pub struct StripePaymentIntentRequest {
...
 
 impl TryFrom<StripePaymentIntentRequest> for payments::PaymentsRequest {
     type Error = error_stack::Report<errors::ApiErrorResponse>;
     fn try_from(item: StripePaymentIntentRequest) -> errors::RouterResult<Self> {
-        Ok(Self {
+        let stripe_mandate = item.payment_method_options.as_ref().and_then(|mandate| {
+            let StripePaymentMethodOptions::Card {
+                request_three_d_secure: _,
+                mandate_options,
+            }: StripePaymentMethodOptions = mandate.clone();
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:159]: try to remove clone here.

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 4 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:522]: done

Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:522]: +1

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 5 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:544]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 6 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:552]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 7 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:565]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 8 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:527]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 10 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:159]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 11]
```diff
@@ -143,12 +144,23 @@ pub struct StripePaymentIntentRequest {
...
     type Error = error_stack::Report<errors::ApiErrorResponse>;
     fn try_from(item: StripePaymentIntentRequest) -> errors::RouterResult<Self> {
-        Ok(Self {
+        let mandate_options_enum = item.payment_method_options.as_ref().and_then(|mandate| {
+            let StripePaymentMethodOptions::Card {
+                request_three_d_secure: _,
+                mandate_options,
+            }: StripePaymentMethodOptions = mandate.clone();
+            mandate_options
+        });
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:160]:
```suggestion
        let (stripe_mandate, authentication_type) = match item.payment_method_options {
            Some(pmo) => {
                let StripePaymentMethodOptions::Card {
                    request_three_d_secure,
                    mandate_options,
                }: StripePaymentMethodOptions = pmo;
                (mandate_options, Some(request_three_d_secure.foreign_into()))
            }
            None => (None, None),
        };
```

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 12]
```diff
@@ -491,15 +509,69 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
 }
 
-#[derive(Default, Eq, PartialEq, Serialize, Deserialize, Clone)]
+#[derive(PartialEq, Eq, Deserialize, Clone, Default, Debug)]
+pub struct MandateOption {
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub accepted_at: Option<time::PrimitiveDateTime>,
+    pub user_agent: Option<String>,
+    pub ip_address: Option<pii::Secret<String, common_utils::pii::IpAddress>>,
+    pub mandate_type: Option<String>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:527]: Can `mandate_type` be an enum

Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:527]: its a string, that we will be passing from euler, so enum is not required

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 11 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs]: can we do this

`let mandate_options_enum = item.payment_method_options.as_ref().and_then(|mandate| {
            let StripePaymentMethodOptions::Card {
                request_three_d_secure: _,
                mandate_options,
            }: StripePaymentMethodOptions = mandate.to_owned();
            mandate_options
        });`

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 13]
```diff
@@ -491,15 +509,69 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+                        acceptance_type: payments::AcceptanceType::Online,
+                        accepted_at: mandate.clone().accepted_at,
+                        online: Some(payments::OnlineMandate {
+                            ip_address: mandate.clone().ip_address.unwrap_or_default(),
+                            user_agent: mandate.to_owned().user_agent.unwrap_or_default(),
+                        }),
+                    },
+                });
+            mandate_data
+        })
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:570]:
```suggestion
let mandate_data = mandate_options.map(|mandate| payments::MandateData {
            mandate_type: if mandate.mandate_type == Some("multi_use".to_string()) {
                payments::MandateType::MultiUse(None)
            } else {
                payments::MandateType::SingleUse(payments::MandateAmountData {
                    amount: mandate.amount.unwrap_or_default(),
                    currency,
                })
            },
            customer_acceptance: payments::CustomerAcceptance {
                acceptance_type: payments::AcceptanceType::Online,
                accepted_at: mandate.accepted_at,
                online: Some(payments::OnlineMandate {
                    ip_address: mandate.ip_address.unwrap_or_default(),
                    user_agent: mandate.user_agent.unwrap_or_default(),
                }),
            },
        });

        Ok(mandate_data)
```

Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:570]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 11 above)
Comment (Abhicodes-crypto) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs]: `to_owned()` and `clone()` both are same in this context.

Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs]: ok

File: crates/router/src/compatibility/stripe/payment_intents/types.rs (hunk 12 above)
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs]: done

**Review by Abhicodes-crypto — APPROVED** [review]

File: crates/router/src/compatibility/stripe/setup_intents/types.rs [hunk 14]
```diff
@@ -108,11 +115,65 @@ impl From<Shipping> for payments::Address {
...
+
+#[derive(Default, Eq, PartialEq, Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum Request3DS {
+    #[default]
+    Automatic,
+    Any,
+}
+
+impl ForeignFrom<Option<Request3DS>> for api_models::enums::AuthenticationType {
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/setup_intents/types.rs:136]: I guess this implementation is there in `payment_intents/types.rs`. Check if that solves the use case here.

**Review by Abhicodes-crypto — DISMISSED** [review]
LGTM. 
Add PR description on what changes are being made so that other reviewers have context.

**Review by Abhicodes-crypto — DISMISSED** [review]

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 15]
```diff
@@ -491,15 +505,83 @@ impl From<payments::PaymentListResponse> for StripePaymentIntentListResponse {
...
+#[derive(Eq, PartialEq, Debug, serde::Deserialize, serde::Serialize, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum StripeMandateType {
+    SingleUse,
+    MultiUse,
+}
+
+#[derive(PartialEq, Eq, Deserialize, Clone, Default, Debug)]
+pub struct MandateOption {
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:526]: yes

[the remark this answers is not in the document]
Comment (Abhicodes-crypto) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:526]: @jagan-jaya , @Narayanbhat166 , there is a deserialisation function to convert unix timestamp format to primitive date time right ?

[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:526]: I could not find any function in datetime module for this deserialization.

[the remark this answers is not in the document]
Comment (Abhicodes-crypto) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:526]: @sahkal  you can use this https://github.com/juspay/hyperswitch/blob/68360d4d6a31d8d7361c83021ca3049780d6d0a3/crates/common_utils/src/custom_serde.rs#L89

**Review by jarnura — CHANGES_REQUESTED** [review]
Build checks are failing

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 16]
```diff
@@ -143,12 +144,26 @@ pub struct StripePaymentIntentRequest {
...
     type Error = error_stack::Report<errors::ApiErrorResponse>;
     fn try_from(item: StripePaymentIntentRequest) -> errors::RouterResult<Self> {
-        Ok(Self {
+        let (mandate_options, authentication_type) = match item.payment_method_options {
+            Some(pmo) => {
+                let StripePaymentMethodOptions::Card {
+                    request_three_d_secure,
+                    mandate_options,
+                }: StripePaymentMethodOptions = pmo;
+                (mandate_options, Some(request_three_d_secure.foreign_into()))
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:160]: done

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 17]
```diff
@@ -143,12 +144,26 @@ pub struct StripePaymentIntentRequest {
...
+            }
+            None => (None, None),
+        };
+        let mandate_data: Option<payments::MandateData> =
+            (mandate_options, item.currency.to_owned()).foreign_try_into()?;
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/payment_intents/types.rs:165]: done

File: crates/router/src/compatibility/stripe/setup_intents.rs [hunk 18]
```diff
@@ -29,7 +29,10 @@ pub async fn setup_intents_create(
         }
     };
 
-    let create_payment_req: payment_types::PaymentsRequest = payload.into();
+    let create_payment_req: payment_types::PaymentsRequest = match payload.try_into() {
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/setup_intents.rs:32]: done

File: crates/router/src/compatibility/stripe/setup_intents.rs [hunk 19]
```diff
@@ -127,7 +130,10 @@ pub async fn setup_intents_update(
         }
     };
 
-    let mut payload: payment_types::PaymentsRequest = stripe_payload.into();
+    let mut payload: payment_types::PaymentsRequest = match stripe_payload.try_into() {
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/setup_intents.rs:133]: done

File: crates/router/src/compatibility/stripe/setup_intents.rs [hunk 20]
```diff
@@ -183,7 +189,10 @@ pub async fn setup_intents_confirm(
...
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/setup_intents.rs:192]: done

File: crates/router/src/compatibility/stripe/setup_intents/types.rs [hunk 21]
```diff
@@ -123,17 +135,40 @@ pub struct StripeSetupIntentRequest {
...
+impl TryFrom<StripeSetupIntentRequest> for payments::PaymentsRequest {
+    type Error = error_stack::Report<errors::ApiErrorResponse>;
+    fn try_from(item: StripeSetupIntentRequest) -> errors::RouterResult<Self> {
+        let (mandate_options, authentication_type) = match item.payment_method_options {
+            Some(pmo) => {
+                let payment_intent::StripePaymentMethodOptions::Card {
+                    request_three_d_secure,
+                    mandate_options,
+                }: payment_intent::StripePaymentMethodOptions = pmo;
+                (mandate_options, Some(request_three_d_secure.foreign_into()))
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/setup_intents/types.rs:152]: done

File: crates/router/src/compatibility/stripe/setup_intents/types.rs [hunk 22]
```diff
@@ -123,17 +135,40 @@ pub struct StripeSetupIntentRequest {
...
+            }
+            None => (None, None),
+        };
+        let mandate_data: Option<payments::MandateData> =
+            (mandate_options, item.currency.to_owned()).foreign_try_into()?;
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/router/src/compatibility/stripe/setup_intents/types.rs:157]: done

**Review by jagan-jaya — APPROVED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]

**Review by jarnura — APPROVED** [review]
